mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp

```diff
@@ -0,0 +1,308 @@
+#include "mlir/Dialect/ArmSME/IR/ArmSME.h"
+#include "mlir/Dialect/ArmSME/Transforms/Passes.h"
+#include "mlir/Dialect/ArmSME/Utils/Utils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/OneToNFuncConversions.h"
+#include "mlir/Dialect/SCF/Transforms/Patterns.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/Transforms/OneToNTypeConversion.h"
+
+#define DEBUG_TYPE "arm-sme-vector-legalization"
+
+namespace mlir::arm_sme {
+#define GEN_PASS_DEF_VECTORLEGALIZATION
+#include "mlir/Dialect/ArmSME/Transforms/Passes.h.inc"
+} // namespace mlir::arm_sme
+
+using namespace mlir;
+using namespace mlir::arm_sme;
+
+namespace {
+
+struct SMETile {
+  // Note: The units of (row, col) are vscale (as SME tiles are scalable).
+  int row{0};
+  int col{0};
+  VectorType type;
+};
+
+/// Adds a constant scalable offset to `indices`. i.e. for 2D:
+/// { indices[0] + offset[0] * vscale, indices[1] + offset[1] * vscale }
+SmallVector<Value, 2> addConstantScalableOffset(OpBuilder &builder,
+                                                Location loc,
+                                                ValueRange indices,
+                                                ArrayRef<int> scalableOffset) {
+  auto vscale = builder.create<vector::VectorScaleOp>(loc);
+  return llvm::map_to_vector(
+      llvm::zip_equal(indices, scalableOffset), [&](auto pair) -> Value {
+        auto [index, base] = pair;
+        auto offset = builder.create<arith::MulIOp>(
+            loc, builder.create<arith::ConstantIndexOp>(loc, base), vscale);
+        return builder.create<arith::AddIOp>(loc, index, offset);
+      });
+}
+
+/// Remaps indices (e.g. from a load/store) for a larger vector type to indices
+/// for one of the SME tiles it will decompose into.
+SmallVector<Value, 2> remapIndicesForSMETile(OpBuilder &builder, Location loc,
```

Comment:
One thing that confuses me ...

Re-mapping indices for a larger vector to indices for a smaller vector should mean that the indices are reduced: a large vector will have a range of e.g. [0:8xVL], smaller vector will have a range of [0:4xVL]. However, it looks like `addConstantScalableOffset` increases the indices rather than "clamping" them.

I think that we are missing some examples :)

Reply:
Decomposing a larger type into smaller types means the indices get bigger, see this diagram: 
```
// Decompose a 8x8 load at [a,b] into four 4x4 loads:
// initial indices = [a,b], size = 8x8
// ┌─────────────┬─────────────┐
// │[a,b]        │[a,b+4]      │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// ├─────────────┼─────────────┤
// │[a+4,b]      │[a+4,b+4]    │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// │             │             │
// └─────────────┴─────────────┘
```

Reply:
I hope you don't think I intentionally named somthing `tileTile` :laughing:, I spent a while renaming things, and messed up a global find and replace :sweat_smile: